File: java/src/main/java/cn/leetcode/stack/$84_LargestRectangleinHistogram.java
package cn.leetcode.stack;

import java.util.Stack;

class $84_LargestRectangleinHistogram {
    public int largestRectangleArea(int[] heights) {
        Stack<Integer> stack = new Stack<>();
        int res=0;
        stack.push(-1);
        for(int i=0;i<heights.length;i++){
            while(stack.peek()!=-1 && heights[stack.peek()]>=heights[i])
                res=Math.max(res , heights[stack.pop()] * (i-stack.peek()-1));
            stack.push(i);
        }

        while(stack.peek()!=-1)
            res=Math.max(res , heights[stack.pop()] * (heights.length-stack.peek()-1));

        return res;
    }
}

File: python/stack/monotone_stack/$84_LargestRectangleInHistogram.py
from typing import List


class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        size = len(heights)
        res = 0

        stack = []

        for i in range(size):
            while stack and heights[i] < heights[stack[-1]]:
                cur_height = heights[stack.pop()]

                while stack and cur_height == heights[stack[-1]]:
                    stack.pop()

                if stack:
                    cur_width = i - stack[-1] - 1
                else:
                    cur_width = i

                res = max(res, cur_height * cur_width)
            stack.append(i)

        while stack:
            cur_height = heights[stack.pop()]
            while stack and cur_height == heights[stack[-1]]:
                stack.pop()

            if stack:
                cur_width = size - stack[-1] - 1
            else:
                cur_width = size
            res = max(res, cur_height * cur_width)

        return res
